fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/bitmap/ComposedBitmapIndexCursor.java

```diff
@@ -0,0 +1,165 @@
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Function;
+
+/**
+ * A {@link RecordCursor} doing a bit-wise merge of bitmaps from two or more {@code BITMAP_VALUE} indexes.
+ *
+ * The bit operations can correspond to a Boolean expression over those indexes' rightmost grouping keys.
+ *
+ * @see BitmapValueIndexMaintainer
+ */
+class ComposedBitmapIndexCursor extends MergeCursor<IndexEntry, IndexEntry, MergeCursorState<IndexEntry>> {
+    @Nonnull
+    private final Composer composer;
+
+    /**
+     * Function for generating a bitmap from several others, all of the same size.
+     */
+    @FunctionalInterface
+    public interface Composer {
+        /**
+         * Generate a bitmap from several others.
+         * @param bitmaps a list of bitmaps or {@code null} if the corresponding input is absent / empty
+         * @param size the common size of the bitmaps
+         * @return a new bitmap formed from the inputs or {@code null} to represent an empty (all zero) bitmap
+         */
+        @Nullable
+        byte[] compose(@Nonnull List<byte[]> bitmaps, int size);
+    }
+
+    protected ComposedBitmapIndexCursor(@Nonnull List<MergeCursorState<IndexEntry>> cursorStates, @Nullable FDBStoreTimer timer, @Nonnull Composer composer) {
+        super(cursorStates, timer);
+        this.composer = composer;
+    }
+
+    @Nonnull
+    @Override
+    protected CompletableFuture<List<MergeCursorState<IndexEntry>>> computeNextResultStates() {
+        final List<MergeCursorState<IndexEntry>> cursorStates = getCursorStates();
+        return whenAll(cursorStates).thenApply(vignore -> {
+            boolean anyHasNext = false;
+            for (MergeCursorState<IndexEntry> cursorState : cursorStates) {
+                if (cursorState.getResult().hasNext()) {
+                    anyHasNext = true;
+                } else if (cursorState.getResult().getNoNextReason().isLimitReached()) {
+                    // Stop if any has reached limit.
+                    return Collections.emptyList();
+                }
+            }
+            if (anyHasNext) {
+                final List<MergeCursorState<IndexEntry>> resultStates = new ArrayList<>();
+                long nextPosition = Long.MAX_VALUE;
+                for (MergeCursorState<IndexEntry> cursorState : cursorStates) {
+                    if (cursorState.getResult().hasNext()) {
+                        final IndexEntry indexEntry = cursorState.getResult().get();
+                        final Tuple indexKey = indexEntry.getKey();
+                        final long position = indexKey.getLong(indexKey.size() - 1);
+                        if (nextPosition > position) {
+                            resultStates.clear();
+                            nextPosition = position;
+                        }
+                        if (nextPosition == position) {
+                            resultStates.add(cursorState);
+                        }
```

Comment:
Hm, this doesn't seem right to me. In particular, I think if you had disjoint ranges of positions for the physical bitmaps from different indexes, then I think it won't merge them together correctly, if I understand how those work. (This doesn't matter for `AND`s, but for `OR`s and `NOT`s, etc.)

Reply:
Did you have a particular failure scenario in mind? In `computeNextResultStates` we should get all the states that share a min next position. Then, in `getNextResult`, we collect the bitmap from those and `null` from the rest. `Composer.compose` treats `null` like an empty bitmap, that is one that is all zeros, that is one that has no positions set.

Reply:
I added a comment for what it's trying to accomplish.

Somewhere along the line I developed a preference for
```
if (x <op> y) {
  x = y;
}
```